test/e2e_node/perf_test.go

```diff
@@ -0,0 +1,332 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e_node
+
+import (
+	"context"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types/filters"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+
+	"github.com/docker/docker/api/types"
+	docker "github.com/docker/docker/client"
+	"github.com/google/uuid"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var (
+	toolboxPodName          = "perf-toolbox-pod"
+	toolboxContainer        = "perf-toolbox"
+	graphingPodName         = "perf-graphing-pod"
+	workloadPodPrefix       = "workload-pod-"
+	workloadContainerPrefix = "workload-container-"
+)
+
+type Workload struct{}
+
+var _ = framework.KubeDescribe("PerfAnalysis [Slow]", func() {
+	f := framework.NewDefaultFramework("perf-test")
+	workload := &Workload{}
+
+	Context("pulling the toolbox image and create the container", func() {
+		BeforeEach(func() {
```

Comment:
Is it possible that we move BeforeEach and AfterEach to test/e2e/framework/perf_gatherer.go?  I am thinking something like test/e2e/framework/resource_usage_gatherer.go that contains the utility functions each e2e_node test can enable through some flag: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L230.

Reply:
The problem is we need to start perf command to record events in a given cgroup, but the cgroup name of workload isn't known until the workload is created. Therefore, we still need to be able to interact with profiling toolbox in the workload context (i.e. start workload and then start perf command).
There may be some workarounds using symlink. I've tried but haven't succeed yet.